src/inc/quic_driver_helpers.h

```diff
@@ -0,0 +1,379 @@
+        }
+        CloseHandle(Overlapped.hEvent);
+        return Error == ERROR_SUCCESS;
+    }
+    bool Run(
+        _In_ uint32_t IoControlCode,
+        _In_ uint32_t TimeoutMs = 30000
+        ) {
+        return Run(IoControlCode, nullptr, 0, TimeoutMs);
+    }
+    template<class T>
+    bool Run(
+        _In_ uint32_t IoControlCode,
+        _In_ const T& Data,
+        _In_ uint32_t TimeoutMs = 30000
+        ) {
+        return Run(IoControlCode, (void*)&Data, sizeof(Data), TimeoutMs);
+    }
+    bool Read(
+        _In_ uint32_t IoControlCode,
+        _Out_writes_bytes_opt_(OutBufferSize)
+            void* OutBuffer,
+        _In_ uint32_t OutBufferSize,
+        _Out_ LPDWORD OutBufferWritten, // Needs to be LPDWORD to match DeviceIoControl
+        _In_ uint32_t TimeoutMs = 30000
+        ) {
+        uint32_t Error;
+        OVERLAPPED Overlapped = { 0 };
+        Overlapped.hEvent = CreateEvent(nullptr, FALSE, FALSE, nullptr);
+        if (!Overlapped.hEvent) {
+            Error = GetLastError();
+            QuicTraceEvent(
+                LibraryErrorStatus,
+                "[ lib] ERROR, %u, %s.",
+                Error,
+                "CreateEvent failed");
+            return false;
+        }
+        QuicTraceLogVerbose(
+            TestSendIoctl,
+            "[test] Sending IOCTL %u.",
+            IoGetFunctionCodeFromCtlCode(IoControlCode));
+        if (!DeviceIoControl(
+                DeviceHandle,
+                IoControlCode,
+                nullptr, 0,
+                OutBuffer, OutBufferSize,
+                OutBufferWritten,
+                &Overlapped)) {
+            Error = GetLastError();
+            if (Error != ERROR_IO_PENDING) {
+                CloseHandle(Overlapped.hEvent);
+                QuicTraceEvent(
+                    LibraryErrorStatus,
+                    "[ lib] ERROR, %u, %s.",
+                    Error,
+                    "DeviceIoControl failed");
+                return false;
+            }
+        }
+        DWORD dwBytesReturned;
+        if (!GetOverlappedResultEx(
+                DeviceHandle,
+                &Overlapped,
+                &dwBytesReturned,
+                TimeoutMs,
+                FALSE)) {
+            Error = GetLastError();
+            if (Error == WAIT_TIMEOUT) {
+                Error = ERROR_TIMEOUT;
+                CancelIoEx(DeviceHandle, &Overlapped);
+            } else {
+                QuicTraceEvent(
+                    LibraryErrorStatus,
+                    "[ lib] ERROR, %u, %s.",
+                    Error,
+                    "GetOverlappedResultEx failed");
+            }
+        } else {
+            Error = ERROR_SUCCESS;
+        }
+        CloseHandle(Overlapped.hEvent);
+        return Error == ERROR_SUCCESS;
+    }
+};
+
+#else
+
+class QuicDriverService {
+public:
+    bool Initialize() { return false; }
+    void Uninitialize() { }
+    bool Start() { return false; }
+};
```

Comment:
I rather not go with this model. I prefer the caller to only reference these classes if _WIN32

Reply:
That requires massive changes to the BVTs

Reply:
Once this is in, I'll do fix this in the BVTs. I'm creating an issue for this and the other issues that  will involve changing the BVTs.